client/push.go

```diff
@@ -156,6 +174,78 @@ func (c *Client) postFile(ctx context.Context, r io.Reader, fileSize int64, imag
 		}
 		return fmt.Errorf("sending file did not succeed: http status code %d", res.StatusCode)
 	}
-
 	return nil
 }
+
+// loggingAdapter is an adapter to redirect log messages from retryablehttp
+// to our logger
+type loggingAdapter struct {
+	logger log.Logger
+}
+
+// Printf implements interface used by retryablehttp
+func (l *loggingAdapter) Printf(fmt string, args ...interface{}) {
+	l.logger.Logf(fmt, args)
+}
+
+// postFileV2 uses V2 API to upload images to SCS library server. This is
+// a three step operation: "create" upload image request, which returns a
+// URL to issue an http PUT operation against, and then finally calls the
+// completion endpoint once upload is complete.
+func (c *Client) postFileV2(ctx context.Context, r io.Reader, fileSize int64, imageID string, callback UploadCallback) error {
+
+	if callback == nil {
+		// fallback to default upload callback
+		callback = &defaultUploadCallback{}
+	}
+
+	postURL := "/v2/imagefile/" + imageID
+	c.Logger.Logf("postFile calling %s", postURL)
+
+	// issue upload request (POST) to obtain presigned S3 URL
+	body := UploadImageRequest{
+		Size: fileSize,
+	}
+
+	objJSON, err := c.apiCreate(ctx, postURL, body)
+	if err != nil {
+		return err
+	}
+
+	var res UploadImageResponse
+	if err := json.Unmarshal(objJSON, &res); err != nil {
+		return nil
+	}
+
+	// set up source file reader
+	callback.InitUpload(fileSize, r)
+
+	// upload (PUT) directly to S3 presigned URL provided above
+	presignedURL := res.Data.UploadURL
+
+	req, err := retryablehttp.NewRequest("PUT", presignedURL, callback.GetReader())
+	if err != nil {
+		return fmt.Errorf("error creating request: %v", err)
+	}
+
+	req.ContentLength = fileSize
+	req.Header.Set("Content-Type", "application/octet-stream")
+
+	// redirect log output from retryablehttp to our logger
+	l := loggingAdapter{
+		logger: c.Logger,
+	}
+
+	client := retryablehttp.NewClient()
+	client.Logger = &l
+
+	_, err = client.Do(req.WithContext(ctx))
+	callback.Finish()
+	if err != nil {
+		return fmt.Errorf("error uploading image: %v", err)
+	}
+
+	// send (PUT) image upload completion
+	_, err = c.apiUpdate(ctx, postURL+"/_complete", UploadImageCompleteRequest{})
```

Comment:
For my own education and understanding, what use a `PUT` here? I would have guessed that a `POST` would have been used but I admit to not knowing all of the reasons for using different methods for different actions.

Reply:
My understanding is that best practice is to use POST to create a resource (in this case, an upload request) and PUT is used to update an existing resource (upload request).